File: src/main/java/aurocosh/divinefavor/common/block/base/ModBlock.java
package aurocosh.divinefavor.common.block.base;

import aurocosh.divinefavor.common.core.ResourceNamer;
import aurocosh.divinefavor.common.item.base.ModItemBlock;
import aurocosh.divinefavor.common.lib.interfaces.IOrdered;
import aurocosh.divinefavor.common.registry.ModRegistries;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.util.ResourceLocation;

public class ModBlock extends Block implements IOrdered {
    private final int orderIndex;

    public ModBlock(String name, Material material, int orderIndex) {
        super(material);
        this.orderIndex = orderIndex;
        ResourceLocation fullName = ResourceNamer.getFullName(name);
        setTranslationKey(fullName.toString());
        setRegistryName(fullName);
        ModRegistries.blocks.register(this);

        ModItemBlock itemBlock = getItemBlock();
        if (itemBlock != null) {
            itemBlock.setRegistryName(fullName);
            ModRegistries.itemBlocks.register(itemBlock);
        }
    }

    @Override
    public int getOrderIndex() {
        return orderIndex;
    }

    protected ModItemBlock getItemBlock() {
        return new ModItemBlock(this, orderIndex);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/block/base/ModBlock.kt
package aurocosh.divinefavor.common.block.base

import aurocosh.divinefavor.common.core.ResourceNamer
import aurocosh.divinefavor.common.item.base.ModItemBlock
import aurocosh.divinefavor.common.lib.interfaces.IOrdered
import aurocosh.divinefavor.common.registry.ModRegistries
import net.minecraft.block.Block
import net.minecraft.block.material.Material

open class ModBlock(name: String, material: Material, override val orderIndex: Int) : Block(material), IOrdered {

    protected open fun getItemBlock(): ModItemBlock? = ModItemBlock(this, orderIndex)

    init {
        val fullName = ResourceNamer.getFullName(name)
        translationKey = fullName.toString()
        registryName = fullName
        ModRegistries.blocks.register(this)

        val itemBlock = getItemBlock()
        if (itemBlock != null) {
            itemBlock.registryName = fullName
            ModRegistries.itemBlocks.register(itemBlock)
        }
    }
}
